File: Tree/124.Binary-Tree-Maximum-Path-Sum/124.Binary-Tree-Maximum-Path-Sum.ts
function maxPathSum(root: TreeNode | null): number {
  let result = Number.MIN_SAFE_INTEGER;
  MaxDownPath(root);
  return result;
  function MaxDownPath(node: TreeNode | null) {
    if (node === null) {
      return 0;
    }
    const leftSum = MaxDownPath(node.left);
    const rightSum = MaxDownPath(node.right);

    let maxTurnSum = node.val;
    if (leftSum >= 0) {
      maxTurnSum += leftSum;
    }
    if (rightSum >= 0) {
      maxTurnSum += rightSum;
    }
    result = Math.max(maxTurnSum, result);

    if (leftSum < 0 && rightSum < 0) {
      return node.val;
    } else {
      return Math.max(leftSum, rightSum) + node.val;
    }
  }
}

File: Tree/124.Binary-Tree-Maximum-Path-Sum/124.Binary-Tree-Maximum-Path-Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        result = float('-inf')
        
        
        def maxDownPathSum(root):
            nonlocal result
            if not root:
                return 0
            
            leftSumMax =  maxDownPathSum(root.left)
            rightSumMax = maxDownPathSum(root.right)
            
            sum = root.val
            if leftSumMax > 0:
                sum += leftSumMax
            if rightSumMax > 0:
                sum += rightSumMax
            result = max(sum, result)
            
            return max(leftSumMax, rightSumMax, 0) + root.val
        
        maxDownPathSum(root)
        return result

        
